casepro/backend/junebug.py

```diff
@@ -463,3 +468,54 @@ def received_junebug_message(request):
         created_on=datetime.now(), has_labels=True)
 
     return JsonResponse(msg.as_json())
+
+
+@csrf_exempt
+def receive_identity_store_optout(request):
+    '''Handles optout notifications from the Identity Store.'''
+    if request.method != 'POST':
+        return JsonResponse({'reason': 'Method not allowed.'}, status=405)
+
+    try:
+        data = json.loads(request.body)
+    except ValueError as e:
+        return JsonResponse(
+            {'reason': 'JSON decode error', 'details': e.message}, status=400)
+
+    try:
+        identity_id = data['identity']
+        optout_type = data['optout_type']
+    except KeyError as e:
+        return JsonResponse(
+            {'reason': 'Both "identity" and "optout_type" must be specified.'},
+            status=400)
+
+    syncer = IdentityStoreContactSyncer()
+
+    # The identity store currently doesn't specify the response format or do
+    # anything with the response.
+
+    org = request.org
+    local_contact = syncer.fetch_local(org, identity_id)
+    if not local_contact:
+        return JsonResponse({
+            'reason': "No Contact for id: " + identity_id}, status=400)
+
+    with local_contact.lock(org, identity_id):
```

Comment:
I'm wondering whether fetching the `local_contact` needs to be inside the `with` block (otherwise someone could modify the contact before we enter the block).